core/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java

```diff
@@ -102,12 +105,12 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
 
     }
 
-    protected abstract ShardIterator shards(ClusterState state, InternalRequest request);
+    protected abstract ShardsIterator shards(ClusterState state, InternalRequest request);
```

Comment:
can we do document what the iteration means? that we will try across groups on failure and that it can be null meaning local execution.